// opencode-diff/dist/ui/widgets.js
import chalk from 'chalk';
export const DEFAULT_WIDGET_STYLES = {
    pending: {
        indicator: '?',
        color: chalk.yellow,
        bgColor: chalk.bgYellow,
    },
    accepted: {
        indicator: '✓',
        color: chalk.green,
        bgColor: chalk.bgGreen,
    },
    rejected: {
        indicator: '✗',
        color: chalk.red,
        bgColor: chalk.bgRed,
    },
    currentLine: {
        indicator: '▶',
        color: chalk.cyan,
    },
};
export class WidgetManager {
    pendingChange = null;
    currentPosition = null;
    styles;
    widgets = new Map();
    visibleWidgets = new Set();
    showCurrentIndicator = true;
    showStateIndicators = true;
    compactMode = false;
    constructor(styles = {}, options = {}) {
        this.styles = {
            pending: { ...DEFAULT_WIDGET_STYLES.pending, ...styles.pending },
            accepted: { ...DEFAULT_WIDGET_STYLES.accepted, ...styles.accepted },
            rejected: { ...DEFAULT_WIDGET_STYLES.rejected, ...styles.rejected },
            currentLine: { ...DEFAULT_WIDGET_STYLES.currentLine, ...styles.currentLine },
        };
        this.showCurrentIndicator = options.showCurrentIndicator ?? true;
        this.showStateIndicators = options.showStateIndicators ?? true;
        this.compactMode = options.compactMode ?? false;
    }
    setPendingChange(change) {
        this.pendingChange = change;
        this.rebuildWidgets();
    }
    setCurrentPosition(position) {
        this.currentPosition = position;
        this.updateCurrentLineIndicator();
    }
    rebuildWidgets() {
        this.widgets.clear();
        this.visibleWidgets.clear();
        if (!this.pendingChange?.parsedDiff) {
            return;
        }
        let absoluteLine = 0;
        const hunks = this.pendingChange.parsedDiff.hunks;
        for (let hunkIndex = 0; hunkIndex < hunks.length; hunkIndex++) {
            const hunk = hunks[hunkIndex];
            for (let lineIndex = 0; lineIndex < hunk.lines.length; lineIndex++) {
                const line = hunk.lines[lineIndex];
                const lineId = `${hunkIndex}:${lineIndex}`;
                const isChangeable = line.type === 'added' || line.type === 'deleted';
                const state = this.pendingChange.getLineState(lineId) || 'pending';
                const widget = {
                    position: {
                        hunkIndex,
                        lineIndex,
                        absoluteLine,
                    },
                    lineId,
                    state,
                    isCurrent: false,
                    isChangeable,
                    lineType: line.type,
                };
                this.widgets.set(lineId, widget);
                if (isChangeable) {
                    this.visibleWidgets.add(lineId);
                }
                absoluteLine++;
            }
            absoluteLine += 2;
        }
    }
    updateCurrentLineIndicator() {
        for (const widget of this.widgets.values()) {
            widget.isCurrent = false;
        }
        if (!this.currentPosition) {
            return;
        }
        const lineId = `${this.currentPosition.hunkIndex}:${this.currentPosition.lineIndex}`;
        const widget = this.widgets.get(lineId);
        if (widget) {
            widget.isCurrent = true;
        }
    }
    getWidgetForLine(hunkIndex, lineIndex) {
        const lineId = `${hunkIndex}:${lineIndex}`;
        return this.widgets.get(lineId);
    }
    getWidgetState(hunkIndex, lineIndex) {
        const widget = this.getWidgetForLine(hunkIndex, lineIndex);
        return widget?.state;
    }
    setWidgetState(hunkIndex, lineIndex, state) {
        const lineId = `${hunkIndex}:${lineIndex}`;
        const widget = this.widgets.get(lineId);
        if (widget) {
            widget.state = state;
        }
    }
    getStateIndicator(state) {
        if (!this.showStateIndicators) {
            return ' ';
        }
        const style = this.styles[state];
        const indicator = this.compactMode ? style.indicator.charAt(0) : style.indicator;
        return style.color(indicator);
    }
    getCurrentLineIndicator() {
        if (!this.showCurrentIndicator) {
            return ' ';
        }
        const style = this.styles.currentLine;
        return style.color(style.indicator);
    }
    renderWidgetForLine(hunkIndex, lineIndex) {
        const widget = this.getWidgetForLine(hunkIndex, lineIndex);
        if (!widget || !widget.isChangeable) {
            return '  ';
        }
        const parts = [];
        if (widget.isCurrent) {
            parts.push(this.getCurrentLineIndicator());
        }
        else {
            parts.push(' ');
        }
        parts.push(this.getStateIndicator(widget.state));
        return parts.join('');
    }
    renderWidgetColumn(hunkIndex, lineIndex, _theme) {
        return this.renderWidgetForLine(hunkIndex, lineIndex);
    }
    getAllVisibleWidgets() {
        return Array.from(this.visibleWidgets)
            .map(id => this.widgets.get(id))
            .filter((w) => w !== undefined);
    }
    getPendingWidgets() {
        return this.getAllVisibleWidgets().filter(w => w.state === 'pending');
    }
    getAcceptedWidgets() {
        return this.getAllVisibleWidgets().filter(w => w.state === 'accepted');
    }
    getRejectedWidgets() {
        return this.getAllVisibleWidgets().filter(w => w.state === 'rejected');
    }
    getWidgetStats() {
        const visible = this.getAllVisibleWidgets();
        return {
            pending: visible.filter(w => w.state === 'pending').length,
            accepted: visible.filter(w => w.state === 'accepted').length,
            rejected: visible.filter(w => w.state === 'rejected').length,
            total: visible.length,
        };
    }
    updateWidgetStatesFromPendingChange() {
        if (!this.pendingChange) {
            return;
        }
        const lineStates = this.pendingChange.getAllLineStates();
        for (const [lineId, state] of lineStates) {
            const widget = this.widgets.get(lineId);
            if (widget) {
                widget.state = state;
            }
        }
    }
    setStyles(styles) {
        this.styles = {
            pending: { ...this.styles.pending, ...styles.pending },
            accepted: { ...this.styles.accepted, ...styles.accepted },
            rejected: { ...this.styles.rejected, ...styles.rejected },
            currentLine: { ...this.styles.currentLine, ...styles.currentLine },
        };
    }
    getStyles() {
        return { ...this.styles };
    }
    setShowCurrentIndicator(show) {
        this.showCurrentIndicator = show;
    }
    setShowStateIndicators(show) {
        this.showStateIndicators = show;
    }
    setCompactMode(compact) {
        this.compactMode = compact;
    }
    reset() {
        this.widgets.clear();
        this.visibleWidgets.clear();
        this.currentPosition = null;
        this.rebuildWidgets();
    }
}
export class AcceptRejectWidget {
    position;
    state;
    onAccept = null;
    onReject = null;
    isHovered = false;
    isFocused = false;
    styles;
    constructor(position, initialState = 'pending', styles = {}) {
        this.position = position;
        this.state = initialState;
        this.styles = {
            pending: { ...DEFAULT_WIDGET_STYLES.pending, ...styles.pending },
            accepted: { ...DEFAULT_WIDGET_STYLES.accepted, ...styles.accepted },
            rejected: { ...DEFAULT_WIDGET_STYLES.rejected, ...styles.rejected },
            currentLine: { ...DEFAULT_WIDGET_STYLES.currentLine, ...styles.currentLine },
        };
    }
    getPosition() {
        return { ...this.position };
    }
    getState() {
        return this.state;
    }
    setState(state) {
        this.state = state;
    }
    accept() {
        this.state = 'accepted';
        this.onAccept?.();
    }
    reject() {
        this.state = 'rejected';
        this.onReject?.();
    }
    setOnAccept(callback) {
        this.onAccept = callback;
    }
    setOnReject(callback) {
        this.onReject = callback;
    }
    setHovered(hovered) {
        this.isHovered = hovered;
    }
    setFocused(focused) {
        this.isFocused = focused;
    }
    render() {
        const style = this.styles[this.state];
        let output = style.color(style.indicator);
        if (this.isFocused) {
            output = chalk.bold(output);
        }
        if (this.isHovered) {
            output = chalk.underline(output);
        }
        return output;
    }
    renderDetailed() {
        const parts = [];
        if (this.isFocused) {
            parts.push(this.styles.currentLine.color('▶'));
        }
        else {
            parts.push(' ');
        }
        parts.push(this.render());
        return parts.join(' ');
    }
}
export class CurrentLineIndicator {
    position = null;
    styles;
    visible = true;
    constructor(styles = {}) {
        this.styles = {
            pending: { ...DEFAULT_WIDGET_STYLES.pending, ...styles.pending },
            accepted: { ...DEFAULT_WIDGET_STYLES.accepted, ...styles.accepted },
            rejected: { ...DEFAULT_WIDGET_STYLES.rejected, ...styles.rejected },
            currentLine: { ...DEFAULT_WIDGET_STYLES.currentLine, ...styles.currentLine },
        };
    }
    setPosition(position) {
        this.position = position;
    }
    getPosition() {
        return this.position ? { ...this.position } : null;
    }
    setVisible(visible) {
        this.visible = visible;
    }
    isVisible() {
        return this.visible;
    }
    render() {
        if (!this.visible || !this.position) {
            return ' ';
        }
        return this.styles.currentLine.color(this.styles.currentLine.indicator);
    }
    renderWithLineNumber(lineNumber) {
        if (!this.visible || !this.position) {
            return String(lineNumber).padStart(4, ' ');
        }
        const indicator = this.styles.currentLine.color(this.styles.currentLine.indicator);
        const numStr = String(lineNumber).padStart(3, ' ');
        return `${indicator}${numStr}`;
    }
}
export function createWidgetManager(pendingChange, styles) {
    const manager = new WidgetManager(styles);
    if (pendingChange) {
        manager.setPendingChange(pendingChange);
    }
    return manager;
}
export function createAcceptRejectWidget(hunkIndex, lineIndex, state = 'pending', styles) {
    return new AcceptRejectWidget({
        hunkIndex,
        lineIndex,
        absoluteLine: 0,
    }, state, styles);
}
export function createCurrentLineIndicator(styles) {
    return new CurrentLineIndicator(styles);
}
export default {
    WidgetManager,
    AcceptRejectWidget,
    CurrentLineIndicator,
    createWidgetManager,
    createAcceptRejectWidget,
    createCurrentLineIndicator,
    DEFAULT_WIDGET_STYLES,
};
