// opencode-diff/dist/config.js
/**
 * Configuration System for OpenCode Diff Plugin
 *
 * Manages plugin configuration loading, validation, and access.
 * Configuration is stored in .opencode/diff-plugin.json
 */
import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'fs';
import { dirname, join } from 'path';
/**
 * Default configuration values
 */
export const DEFAULT_CONFIG = {
    enabled: true,
    autoAccept: [],
    autoReject: [],
    maxFileSize: 1024 * 1024, // 1MB
    theme: 'auto',
    showLineNumbers: true,
    confirmRejectAll: true,
    keybindings: [],
    ide: {
        enabled: false,
        stateFilePath: '.opencode/.diff-plugin-state.json',
    },
};
/**
 * Configuration validation error
 */
export class ConfigValidationError extends Error {
    path;
    constructor(message, path) {
        super(message);
        this.path = path;
        this.name = 'ConfigValidationError';
    }
}
/**
 * Configuration Manager
 *
 * Handles loading, saving, and validating plugin configuration.
 * Uses .opencode/diff-plugin.json as the config file.
 */
export class ConfigManager {
    config;
    configPath;
    workspaceRoot;
    /**
     * Create a new ConfigManager
     * @param workspaceRoot - The workspace root directory
     */
    constructor(workspaceRoot) {
        this.workspaceRoot = workspaceRoot;
        this.configPath = join(workspaceRoot, '.opencode', 'diff-plugin.json');
        this.config = this.load();
    }
    /**
     * Load configuration from file or use defaults
     */
    load() {
        if (!existsSync(this.configPath)) {
            return { ...DEFAULT_CONFIG };
        }
        try {
            const content = readFileSync(this.configPath, 'utf-8');
            const parsed = JSON.parse(content);
            return this.mergeWithDefaults(parsed);
        }
        catch (error) {
            console.warn(`[DiffPlugin] Failed to load config from ${this.configPath}:`, error instanceof Error ? error.message : String(error));
            return { ...DEFAULT_CONFIG };
        }
    }
    /**
     * Merge loaded config with defaults
     */
    mergeWithDefaults(loaded) {
        return {
            enabled: loaded.enabled ?? DEFAULT_CONFIG.enabled,
            autoAccept: loaded.autoAccept ?? [...DEFAULT_CONFIG.autoAccept],
            autoReject: loaded.autoReject ?? [...DEFAULT_CONFIG.autoReject],
            maxFileSize: loaded.maxFileSize ?? DEFAULT_CONFIG.maxFileSize,
            theme: loaded.theme ?? DEFAULT_CONFIG.theme,
            showLineNumbers: loaded.showLineNumbers ?? DEFAULT_CONFIG.showLineNumbers,
            confirmRejectAll: loaded.confirmRejectAll ?? DEFAULT_CONFIG.confirmRejectAll,
            keybindings: loaded.keybindings ?? [...DEFAULT_CONFIG.keybindings],
            ide: {
                enabled: loaded.ide?.enabled ?? DEFAULT_CONFIG.ide.enabled,
                stateFilePath: loaded.ide?.stateFilePath ?? DEFAULT_CONFIG.ide.stateFilePath,
            },
        };
    }
    /**
     * Validate the current configuration
     */
    validate() {
        const errors = [];
        // Validate enabled
        if (typeof this.config.enabled !== 'boolean') {
            errors.push('enabled must be a boolean');
        }
        // Validate autoAccept
        if (!Array.isArray(this.config.autoAccept)) {
            errors.push('autoAccept must be an array');
        }
        else {
            for (let i = 0; i < this.config.autoAccept.length; i++) {
                if (typeof this.config.autoAccept[i] !== 'string') {
                    errors.push(`autoAccept[${i}] must be a string`);
                }
            }
        }
        // Validate autoReject
        if (!Array.isArray(this.config.autoReject)) {
            errors.push('autoReject must be an array');
        }
        else {
            for (let i = 0; i < this.config.autoReject.length; i++) {
                if (typeof this.config.autoReject[i] !== 'string') {
                    errors.push(`autoReject[${i}] must be a string`);
                }
            }
        }
        // Validate maxFileSize
        if (typeof this.config.maxFileSize !== 'number' || this.config.maxFileSize < 0) {
            errors.push('maxFileSize must be a non-negative number');
        }
        // Validate theme
        if (!['light', 'dark', 'auto'].includes(this.config.theme)) {
            errors.push('theme must be "light", "dark", or "auto"');
        }
        // Validate showLineNumbers
        if (typeof this.config.showLineNumbers !== 'boolean') {
            errors.push('showLineNumbers must be a boolean');
        }
        // Validate confirmRejectAll
        if (typeof this.config.confirmRejectAll !== 'boolean') {
            errors.push('confirmRejectAll must be a boolean');
        }
        // Validate keybindings
        if (!Array.isArray(this.config.keybindings)) {
            errors.push('keybindings must be an array');
        }
        else {
            const validActions = [
                'acceptLine', 'rejectLine', 'acceptHunk', 'rejectHunk',
                'acceptFile', 'rejectFile', 'acceptAll', 'rejectAll',
                'nextLine', 'prevLine', 'nextHunk', 'prevHunk',
                'nextFile', 'prevFile', 'scrollUp', 'scrollDown',
                'pageUp', 'pageDown', 'goToTop', 'goToBottom',
                'toggleHelp', 'quit', 'undo', 'redo'
            ];
            for (let i = 0; i < this.config.keybindings.length; i++) {
                const kb = this.config.keybindings[i];
                if (!kb || typeof kb !== 'object') {
                    errors.push(`keybindings[${i}] must be an object`);
                    continue;
                }
                if (typeof kb.key !== 'string' || kb.key.length === 0) {
                    errors.push(`keybindings[${i}].key must be a non-empty string`);
                }
                if (!validActions.includes(kb.action)) {
                    errors.push(`keybindings[${i}].action must be a valid KeyboardAction`);
                }
            }
        }
        // Validate ide config
        if (this.config.ide !== undefined) {
            if (typeof this.config.ide !== 'object' || this.config.ide === null) {
                errors.push('ide must be an object');
            }
            else {
                if (this.config.ide.enabled !== undefined && typeof this.config.ide.enabled !== 'boolean') {
                    errors.push('ide.enabled must be a boolean');
                }
                if (this.config.ide.stateFilePath !== undefined && typeof this.config.ide.stateFilePath !== 'string') {
                    errors.push('ide.stateFilePath must be a string');
                }
            }
        }
        return {
            valid: errors.length === 0,
            errors,
        };
    }
    /**
     * Validate configuration and throw on error
     */
    validateOrThrow() {
        const result = this.validate();
        if (!result.valid) {
            throw new ConfigValidationError(`Configuration validation failed: ${result.errors.join(', ')}`, this.configPath);
        }
    }
    /**
     * Save current configuration to file
     */
    save() {
        try {
            const dir = dirname(this.configPath);
            if (!existsSync(dir)) {
                mkdirSync(dir, { recursive: true });
            }
            writeFileSync(this.configPath, JSON.stringify(this.config, null, 2), 'utf-8');
        }
        catch (error) {
            throw new Error(`Failed to save config to ${this.configPath}: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    /**
     * Reload configuration from file
     */
    reload() {
        this.config = this.load();
    }
    /**
     * Update configuration with partial values
     */
    update(updates) {
        this.config = { ...this.config, ...updates };
    }
    /**
     * Reset configuration to defaults
     */
    reset() {
        this.config = { ...DEFAULT_CONFIG };
    }
    // Getter methods
    /**
     * Check if plugin is enabled
     */
    isEnabled() {
        return this.config.enabled;
    }
    /**
     * Get auto-accept glob patterns
     */
    getAutoAcceptPatterns() {
        return [...this.config.autoAccept];
    }
    /**
     * Get auto-reject glob patterns
     */
    getAutoRejectPatterns() {
        return [...this.config.autoReject];
    }
    /**
     * Get maximum file size in bytes
     */
    getMaxFileSize() {
        return this.config.maxFileSize;
    }
    /**
     * Get theme preference
     */
    getTheme() {
        return this.config.theme;
    }
    /**
     * Check if line numbers should be shown
     */
    shouldShowLineNumbers() {
        return this.config.showLineNumbers;
    }
    /**
     * Check if confirmation is required before rejecting all
     */
    shouldConfirmRejectAll() {
        return this.config.confirmRejectAll;
    }
    /**
     * Get custom keybindings
     */
    getKeybindings() {
        return [...this.config.keybindings];
    }
    /**
     * Get the full configuration object
     */
    getConfig() {
        return { ...this.config };
    }
    /**
     * Get the configuration file path
     */
    getConfigPath() {
        return this.configPath;
    }
    /**
     * Check if a file matches any auto-accept pattern
     */
    shouldAutoAccept(filePath) {
        return this.matchGlobPatterns(filePath, this.config.autoAccept);
    }
    /**
     * Check if a file matches any auto-reject pattern
     */
    shouldAutoReject(filePath) {
        return this.matchGlobPatterns(filePath, this.config.autoReject);
    }
    /**
     * Check if a file exceeds the maximum file size
     */
    isFileTooLarge(fileSize) {
        if (this.config.maxFileSize === 0) {
            return false;
        }
        return fileSize > this.config.maxFileSize;
    }
    /**
     * Match a file path against glob patterns
     * Supports basic glob patterns: *, ?, **
     */
    matchGlobPatterns(filePath, patterns) {
        for (const pattern of patterns) {
            if (this.matchGlob(filePath, pattern)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Match a file path against a single glob pattern
     */
    matchGlob(filePath, pattern) {
        let regexPattern = pattern;
        regexPattern = regexPattern
            .replace(/\*\*\//g, '__GLOBSTAR_SLASH__')
            .replace(/\/\*\*/g, '__SLASH_GLOBSTAR__')
            .replace(/\*\*/g, '__GLOBSTAR__')
            .replace(/\*/g, '__STAR__')
            .replace(/\?/g, '__QUESTION__')
            .replace(/__GLOBSTAR_SLASH__/g, '(?:.*/)?')
            .replace(/__SLASH_GLOBSTAR__/g, '(?:/.*)?')
            .replace(/__GLOBSTAR__/g, '.*')
            .replace(/__STAR__/g, '[^/]*')
            .replace(/__QUESTION__/g, '.');
        const regex = new RegExp(`^${regexPattern}$`);
        return regex.test(filePath);
    }
}
export default ConfigManager;
