// opencode-diff/dist/debug.js
/**
 * Debug logging utility for OpenCode Diff Plugin
 *
 * Provides leveled logging with environment variable control
 */
export class DebugLogger {
    static instance;
    logLevel;
    component;
    logs = [];
    maxLogs = 1000;
    constructor(component) {
        this.component = component;
        this.logLevel = this.detectLogLevel();
    }
    static getInstance(component) {
        if (!DebugLogger.instance) {
            DebugLogger.instance = new DebugLogger(component);
        }
        return new DebugLogger(component);
    }
    detectLogLevel() {
        const envLevel = process.env.DEBUG?.toLowerCase();
        if (envLevel?.includes('opencode-diff-plugin') || envLevel === '*') {
            return 'debug';
        }
        if (envLevel?.includes('error'))
            return 'error';
        if (envLevel?.includes('warn'))
            return 'warn';
        if (envLevel?.includes('info'))
            return 'info';
        return 'info';
    }
    setLogLevel(level) {
        this.logLevel = level;
    }
    shouldLog(level) {
        const levels = ['debug', 'info', 'warn', 'error'];
        return levels.indexOf(level) >= levels.indexOf(this.logLevel);
    }
    formatMessage(level, message) {
        return `[${this.component}] [${level.toUpperCase()}] ${message}`;
    }
    addToHistory(entry) {
        this.logs.push(entry);
        if (this.logs.length > this.maxLogs) {
            this.logs.shift();
        }
    }
    debug(message, data) {
        if (!this.shouldLog('debug'))
            return;
        const entry = {
            timestamp: new Date().toISOString(),
            level: 'debug',
            component: this.component,
            message,
            data
        };
        this.addToHistory(entry);
        console.log(this.formatMessage('debug', message), data !== undefined ? data : '');
    }
    info(message, data) {
        if (!this.shouldLog('info'))
            return;
        const entry = {
            timestamp: new Date().toISOString(),
            level: 'info',
            component: this.component,
            message,
            data
        };
        this.addToHistory(entry);
        console.log(this.formatMessage('info', message), data !== undefined ? data : '');
    }
    warn(message, data) {
        if (!this.shouldLog('warn'))
            return;
        const entry = {
            timestamp: new Date().toISOString(),
            level: 'warn',
            component: this.component,
            message,
            data
        };
        this.addToHistory(entry);
        console.warn(this.formatMessage('warn', message), data !== undefined ? data : '');
    }
    error(message, error) {
        if (!this.shouldLog('error'))
            return;
        const entry = {
            timestamp: new Date().toISOString(),
            level: 'error',
            component: this.component,
            message,
            data: error instanceof Error ? { message: error.message, stack: error.stack } : error
        };
        this.addToHistory(entry);
        console.error(this.formatMessage('error', message), error !== undefined ? error : '');
    }
    group(label) {
        if (this.shouldLog('debug')) {
            console.group(this.formatMessage('debug', label));
        }
    }
    groupEnd() {
        if (this.shouldLog('debug')) {
            console.groupEnd();
        }
    }
    getLogs() {
        return [...this.logs];
    }
    clearLogs() {
        this.logs = [];
    }
    exportLogs() {
        return JSON.stringify(this.logs, null, 2);
    }
}
export function createLogger(component) {
    return DebugLogger.getInstance(component);
}
export function enableDebugMode() {
    process.env.DEBUG = 'opencode-diff-plugin*';
    console.log('[Debug] Debug mode enabled. Set DEBUG=opencode-diff-plugin* to see all logs.');
}
export function disableDebugMode() {
    delete process.env.DEBUG;
    console.log('[Debug] Debug mode disabled.');
}
