// opencode-diff/dist/interceptor.js
import { ChangeQueue, changeQueue, InterceptedError, PendingChange } from './state-manager.js';
import { createLogger } from './debug.js';
const logger = createLogger('Interceptor');
export class ToolInterceptor {
    context;
    constructor(context) {
        this.context = context;
    }
    async before(tool, args) {
        if (tool !== 'write' && tool !== 'edit') {
            logger.debug('Skipping non-intercepted tool', { tool });
            return;
        }
        const { filePath } = this.extractFilePath(tool, args);
        logger.info('Intercepting tool call', { tool, filePath, callID: this.context.callID });
        const oldContent = await this.readOriginalContent(filePath);
        const newContent = this.computeNewContent(tool, oldContent, args);
        const isNewFile = oldContent === '';
        logger.debug('Content computed', {
            filePath,
            isNewFile,
            oldContentLength: oldContent.length,
            newContentLength: newContent.length,
        });
        const change = new PendingChange({
            id: ChangeQueue.generateId(),
            tool,
            filePath,
            oldContent,
            newContent,
            sessionID: this.context.sessionID,
            callID: this.context.callID,
            timestamp: Date.now(),
        });
        this.context.changeQueue.add(change);
        const pendingCount = this.context.changeQueue.size();
        logger.info('Change queued successfully', {
            changeId: change.id,
            filePath,
            pendingCount,
        });
        throw new InterceptedError(`Tool '${tool}' for '${filePath}' intercepted by DiffPlugin`, change.id, filePath);
    }
    async after(tool) {
        if (tool !== 'write' && tool !== 'edit') {
            return;
        }
        console.log(`[DiffPlugin] Tool '${tool}' executed successfully`);
    }
    extractFilePath(tool, args) {
        if (typeof args !== 'object' || args === null) {
            throw new Error(`Invalid arguments for ${tool} tool: expected object`);
        }
        const argsObj = args;
        const filePath = argsObj.filePath;
        if (typeof filePath !== 'string') {
            throw new Error(`Invalid filePath for ${tool} tool: expected string`);
        }
        return { filePath };
    }
    async readOriginalContent(filePath) {
        try {
            const fullPath = this.resolvePath(filePath);
            const result = await this.context.$ `cat ${fullPath}`;
            return result.stdout.toString();
        }
        catch {
            return '';
        }
    }
    computeNewContent(tool, oldContent, args) {
        if (tool === 'write') {
            const writeArgs = args;
            return writeArgs.content;
        }
        if (tool === 'edit') {
            const editArgs = args;
            return oldContent.replace(editArgs.oldString, editArgs.newString);
        }
        return oldContent;
    }
    resolvePath(filePath) {
        if (filePath.startsWith('/')) {
            return filePath;
        }
        return `${this.context.directory}/${filePath}`;
    }
}
export function createBeforeHandler($, directory, _configManager, queue = changeQueue) {
    return async (input, output) => {
        const interceptor = new ToolInterceptor({
            $,
            directory,
            sessionID: input.sessionID,
            callID: input.callID,
            changeQueue: queue,
        });
        await interceptor.before(input.tool, output.args);
    };
}
export function createAfterHandler($, directory, _configManager, queue = changeQueue) {
    return async (input) => {
        const interceptor = new ToolInterceptor({
            $,
            directory,
            sessionID: input.sessionID,
            callID: input.callID,
            changeQueue: queue,
        });
        await interceptor.after(input.tool);
    };
}
